File: App_Resources/Android/src/main/java/com/faceclaw/app/g2protocol/OutboundMessage.java
package com.faceclaw.app;
import java.util.Arrays;

public final class OutboundMessage {
    final String kind;
    final String label;
    final int sid;
    final int flag;
    int magic;
    final byte[] message;
    final int ackTimeoutMs;
    final int tileIndex;
    final boolean isLeftArmMessage;

    int cfwAckLenses, cfwRetries;
    boolean cfwRetryPending;
    final int cfwChecksum;

    int imageUpdateId;
    int imageMessageNumber;
    int imageMessageCount;
    long sentAtMs;
    long writeStartedAtMs;
    long ackDeadlineAtMs;
    byte[] ackPayload = new byte[0];
    Runnable onSent;
    Runnable onAck;
    Runnable onTimeout;

    OutboundMessage(String kind, String label, int sid, int flag, int magic, byte[] message, int ackTimeoutMs, int tileIndex, boolean isLeftArmMessage) {
        this.kind = kind;
        this.label = label;
        this.sid = sid;
        this.flag = flag;
        this.magic = magic;
        this.message = message == null ? new byte[0] : Arrays.copyOf(message, message.length);
        this.cfwChecksum = sid == CfwTransport.SID ? CfwTransport.crc(this.message) : 0;
        this.ackTimeoutMs = ackTimeoutMs;
        this.tileIndex = tileIndex;
        this.isLeftArmMessage = isLeftArmMessage;
    }

    /** NACK names this attempt's stream/message sequence, even without a decoded size. */
    boolean acceptCfwAck(CfwTransport.Ack ack) {
        if (ack == null || sid != CfwTransport.SID || magic != ack.streamId
                || ack.messageId != 0 || (ack.lens != 1 && ack.lens != 2)) return false;
        if (ack.nack) {
            cfwRetryPending = true;
            return false;
        }
        if (message.length != ack.size || cfwChecksum != ack.checksum) return false;
        cfwAckLenses |= ack.lens;
        return !cfwRetryPending && cfwAckLenses == CfwTransport.BOTH;
    }

    /** Replace an attempt atomically under the communicator lock. */
    void prepareCfwReplay(int streamId) {
        magic = streamId;
        ++cfwRetries;
        cfwAckLenses = 0;
        cfwRetryPending = false;
        ackDeadlineAtMs = 0;
        ackPayload = new byte[0];
    }

    void setImageUpdatePosition(int updateId, int messageNumber, int messageCount) {
        this.imageUpdateId = updateId;
        this.imageMessageNumber = messageNumber;
        this.imageMessageCount = messageCount;
    }
}

File: native/kotlin/shared/src/commonMain/kotlin/com/faceclaw/app/g2protocol/OutboundMessage.kt
package com.faceclaw.app

import kotlin.jvm.JvmField

/** Mutable attempt state, owned by the communicator and guarded by its lock. */
class OutboundMessage(
    @JvmField val kind: String?,
    @JvmField val label: String?,
    @JvmField val sid: Int,
    @JvmField val flag: Int,
    @JvmField var magic: Int,
    message: ByteArray?,
    @JvmField val ackTimeoutMs: Int,
    @JvmField val tileIndex: Int,
    @JvmField val isLeftArmMessage: Boolean,
) {
    @JvmField val message: ByteArray = message?.copyOf() ?: ByteArray(0)
    @JvmField
    val cfwChecksum: Int = if (sid == CfwTransport.SID) CfwTransport.crc(this.message) else 0
    @JvmField var cfwAckLenses = 0
    @JvmField var cfwRetries = 0
    @JvmField var cfwRetryPending = false
    @JvmField var imageUpdateId = 0
    @JvmField var imageMessageNumber = 0
    @JvmField var imageMessageCount = 0
    @JvmField var sentAtMs = 0L
    @JvmField var writeStartedAtMs = 0L
    @JvmField var ackDeadlineAtMs = 0L
    @JvmField var ackPayload = ByteArray(0)
    @JvmField var onSent: MessageCallback? = null
    @JvmField var onAck: MessageCallback? = null
    @JvmField var onTimeout: MessageCallback? = null

    /** NACK names this attempt's sequence, even without a decoded size. */
    fun acceptCfwAck(ack: CfwTransport.Ack?): Boolean {
        if (
            ack == null ||
                sid != CfwTransport.SID ||
                magic != ack.streamId ||
                ack.messageId != 0 ||
                (ack.lens != 1 && ack.lens != 2)
        )
            return false
        if (ack.nack) {
            cfwRetryPending = true
            return false
        }
        if (message.size != ack.size || cfwChecksum != ack.checksum) return false
        cfwAckLenses = cfwAckLenses or ack.lens
        return !cfwRetryPending && cfwAckLenses == CfwTransport.BOTH
    }

    /** Replace an attempt atomically under the communicator lock. */
    fun prepareCfwReplay(streamId: Int) {
        magic = streamId
        cfwRetries++
        cfwAckLenses = 0
        cfwRetryPending = false
        ackDeadlineAtMs = 0
        ackPayload = ByteArray(0)
    }

    fun setImageUpdatePosition(updateId: Int, messageNumber: Int, messageCount: Int) {
        imageUpdateId = updateId
        imageMessageNumber = messageNumber
        imageMessageCount = messageCount
    }
}
